vendor/WEB_CONCURRENCY.sh

```diff
@@ -1,29 +1,78 @@
-32768)
-  export DYNO_RAM=14336
-  export WEB_CONCURRENCY=${WEB_CONCURRENCY:-11}
-  ;;
-
-esac
+#!/usr/bin/env bash
+
+# This script was created by the Python buildpack to automatically set the `WEB_CONCURRENCY`
+# environment variable at dyno boot (if it's not already set), based on the available memory
+# and number of CPU cores. The env var is then used by some Python web servers (such as
+# gunicorn and uvicorn) to control the default number of server processes that they launch.
+#
+# The default `WEB_CONCURRENCY` value is calculated as the lowest of either:
+# - `<number of dyno CPU cores> * 2 + 1`
+# - `<dyno available RAM in MB> / 256` (to ensure each process has at least 256 MB RAM)
+#
+# Currently, on Heroku dynos this results in the following concurrency values:
+# - Eco / Basic / Standard-1X: 2 (capped by the 512 MB available memory)
+# - Standard-2X / Private-S / Shield-S: 4 (capped by the 1 GB available memory)
+# - Performance-M / Private-M / Shield-M: 5 (based on the 2 CPU cores)
+# - Performance-L / Private-L / Shield-L: 17 (based on the 8 CPU cores)
+#
+# To override these default values, either set `WEB_CONCURRENCY` as an explicit config var
+# on the app, or pass `--workers <num>` when invoking gunicorn/uvicorn in your Procfile.
+
+function detect_memory_limit_in_mb() {
+  local memory_limit_file='/sys/fs/cgroup/memory/memory.limit_in_bytes'
+
+  # This memory limits file only exists on Heroku, or when using cgroups v1 (Docker < 20.10).
+  if [[ -f "${memory_limit_file}" ]]; then
+    local memory_limit_in_mb=$(($(cat "${memory_limit_file}") / 1048576))
+
+    # Ignore values above 1TB RAM, since when using cgroups v1 the limits file reports a
+    # bogus value of thousands of TB RAM when there is no container memory limit set.
+    if ((memory_limit_in_mb <= 1048576)); then
+      echo "${memory_limit_in_mb}"
+    fi
+  fi
+}
+
+function output() {
+  # Only display log output for web dynos, to prevent breaking one-off dyno scripting use-cases,
+  # and to prevent confusion from messages about WEB_CONCURRENCY in the logs of non-web workers.
+  # (We still actually set the env vars for all dyno types for consistency and easier debugging.)
+  if [[ "${DYNO:-}" == web.* ]]; then
+    echo "Python buildpack: $*" >&2
+  fi
+}
+
+available_memory_in_mb=$(detect_memory_limit_in_mb)
+minimum_memory_per_process_in_mb=256
+
+if [[ -z "${available_memory_in_mb}" ]]; then
+  output "Skipping automatic configuration of WEB_CONCURRENCY since unable to determine available memory."
+  return 0
+fi
+
+# Prevents WEB_CONCURRENCY being set to zero if the environment is extremely memory constrained.
+if ((available_memory_in_mb < minimum_memory_per_process_in_mb)); then
+  max_concurrency_for_available_memory=1
+else
+  max_concurrency_for_available_memory=$((available_memory_in_mb / minimum_memory_per_process_in_mb))
+fi
+
+# This buildpack only supports being run on Heroku's base images, which all have nproc installed.
+cpu_cores=$(nproc)
```

Comment:
Should this handle cases where `nproc` isn't there? Maybe fall back to `getconf -a | grep -E '_?NPROCESSORS_ONLN' | head -n1 | tr -s ' ' | cut -d" " -f2`?

Reply:
Alternatively, you could test this in a `heroku run` by prefixing your command with `nproc() { echo ''; }; export -f getconf;` if you wanted to.

Reply:
As the comment says (the comment was added because I knew you'd ask 😉), this buildpack only supports being run on our base images, and `nproc` is installed in all of them, so if someone tries running it in another image then they are well out of the supported use-case. 

Whilst we could technically have a `command -v nproc` conditional here, I decided against it since I felt we need to draw the line somewhere about what we can expect of the environment, given the buildpack is only intended for use on our images (and consumers like Dokku do use our images). As additional datapoints, I see the PHP buildpack doesn't check to see whether `python` is available [here](https://github.com/heroku/heroku-buildpack-php/blob/752a82c057408338293f25f4c423d53928390507/support/build/apache#L77) and several other buildpacks don't check whether `jq` is available etc (which similarly seem like reasonable assumptions, given that both `python` and `jq` are available in our base images).

That said, for completeness I had tested locally what the experience was when `nproc` was missing, and it results in `cpu_cores` being the empty string, which results in `WEB_CONCURRENCY=1` since the empty string is treated as zero in `$((cpu_cores * 2 + 1))`. So it at least falls back to something safe, and there will be clues for people to follow due to the `nproc: command not found` printed to stderr.

It's also worth noting that most non-Heroku environments won't even get as far as the `nproc` call, since they won't have the `/sys/fs/cgroup/memory/memory.limit_in_bytes` file, and so will have bailed earlier with the message:

```
Python buildpack: Skipping automatic configuration of WEB_CONCURRENCY since unable to determine available memory.
```

Reply:
I've decided to add an exit code check for the `nproc` call in daac379d63d493052600be2e152ccafab138c555, to cover some unknown future case where nproc can't calculate the number of available CPUs (perhaps because of some platform misconfiguration), which has the added bonus side-effect of covering the "nproc missing since the buildpack is being run on an unsupported base image" case too :-)